### PR TITLE
Cleanup Cargo registry data folders

### PR DESCRIPTION
**Pull request by vmapetr**
# Description
This PR add cleanup of Cargo cached registry data folders `Index` and `Src` in Rust provisioners for Windows and Linux images.

#### Related issue:
[Old cargo cache left over from the build in Windows and Linux #895](https://github.com/actions/virtual-environments/issues/895) 

## Check list
- [x] Related issue / work item is attached
- [ ] Changes are tested and related VM images are successfully generated

### CODE REVIEW

File: images/win/scripts/Installers/Install-Rust.ps1 [hunk 1]
```diff
@@ -44,5 +44,10 @@ if exist $env:RUSTUP_HOME (
 $cmdPath = "C:\Rust\rustsym.bat"
 $cmdRustSymScript | Out-File -Encoding ascii -FilePath $cmdPath
 
+# Cleanup Cargo crates cache
+$cargoRegistryPath = Join-Path $env:CARGO_HOME "registry"
+Remove-Item "$cargoRegistryPath/index/*" -Recurse -Force
```
Comment (maxim-lobanov) [inline images/win/scripts/Installers/Install-Rust.ps1:49]: Use `Windows`-style paths `$cargoRegistryPath\index\*`

File: images/linux/scripts/installers/rust.sh [hunk 2]
```diff
@@ -32,6 +32,10 @@ for cmd in rustup rustc rustdoc cargo rustfmt cargo-clippy bindgen cbindgen; do
     fi
 done
 
+# Cleanup Cargo cache
+rm -rf "${CARGO_HOME}/registry/index/*"
+rm -rf "${CARGO_HOME}/registry/src/*"
```
Comment (mzabaluev) [inline images/linux/scripts/installers/rust.sh:37]: I think this should remove the entire `${CARGO_HOME}/registry`.
`${CARGO_HOME}/registry/cache` will only contain crate archives that happened to be downloaded to build and install `bindgen`/`cbindgen`. These can be only occasionally useful in crate builds in a workflow, which may well use different versions of the same crates depending on what their `Cargo.lock` has settled on.
I think it's better to clean these up entirely from the virtual environment and direct the users to [caching schemes](https://github.com/actions/cache/blob/master/examples.md#rust---cargo).

**Comment by mzabaluev** [comment]
Please be aware that removing a pre-populated `~/.cargo/registry/index` may make cargo jobs run slower due to the need to clone the index repository's master branch from scratch. However, when the baked-in index checkout gets too old, it may actually run slower on Windows than a clean clone. The maintainers of `crates.io-index` also periodically reroot the master branch on a squashed commit, so a single-branch clone or fetch performed by cargo should not get exceedingly large.

But again, the right answer to speeding up cargo index sync is caching.

File: images/linux/scripts/installers/rust.sh (hunk 2 above)
Comment (vmapetr) [reply images/linux/scripts/installers/rust.sh]: Thanks!

**Comment by mzabaluev** [comment]
[This workflow run](https://github.com/mzabaluev/github-actions-playground/actions/runs/111944173) simulates the effects of the cleanup on cargo build speeds, against a caching scheme that has all dependencies for the build restored prior to running cargo. The crate's `Cargo.lock` lists 71 dependencies, a medium-sized amount. There is surprisingly little difference in build speeds. Furthermore, on Windows the cache-assisted build actually takes longer. This is probably a bug in cargo.

**Comment by mzabaluev** [comment]
> There is surprisingly little difference in build speeds. Furthermore, on Windows the cache-assisted build actually takes longer.

[Solved this](https://github.com/mzabaluev/github-actions-playground/actions/runs/112187488) by adding the `--offline` flag to build commands.

**Comment by mzabaluev** [comment]
I've got [a test case](https://github.com/mzabaluev/github-actions-playground/blob/ccbdc0e6d1a649516b158575095777fa1b18b422/.github/workflows/test-tar.yml) that can be used to verify the fix.

**Comment by mzabaluev** [comment]
Here's [a workflow](https://github.com/mzabaluev/github-actions-playground/blob/1f6d9fbab2699000ef9a51015e08ce41c15418b4/.github/workflows/test-tar.yml) to verify the fix with. It's poorly suitable for a regression test though, as it spends a lot of time on cargo syncing the index over the network.

**Pull request merged**